pkg/controller/bastion/actuator_delete.go

```diff
@@ -0,0 +1,199 @@
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "delete")
+
+	var factory = azureclient.NewAzureClientFactory(a.client)
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return err
+	}
+
+	err = removeBastionInstance(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove bastion instance: %w", err)
+	}
+
+	deleted, err := isInstanceDeleted(ctx, factory, opt)
+	if err != nil {
+		return fmt.Errorf("failed to check for bastion instance: %w", err)
+	}
+
+	if !deleted {
+		return &ctrlerror.RequeueAfterError{
+			RequeueAfter: 10 * time.Second,
+			Cause:        errors.New("bastion instance is still deleting"),
+		}
+	}
+
+	err = removeNic(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove nic: %w", err)
+	}
+
+	err = removePublicIP(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove public ip: %w", err)
+	}
+
+	err = removeDisk(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove disk: %w", err)
+	}
+
+	err = removeNSGRule(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove nsg rules: %w", err)
+	}
+
+	return nil
+}
+
+func removeNSGRule(ctx context.Context, factory azureclient.Factory, logger logr.Logger, opt *Options) error {
+	securityGroupResp, err := getNetworkSecurityGroup(ctx, factory, opt)
+	if err != nil {
+		return err
+	}
+
+	ruleList := []string{NSGIngressAllowSSHResourceName(opt.BastionInstanceName), NSGEgressDenyAllResourceName(opt.BastionInstanceName), NSGEgressAllowOnlyResourceName(opt.BastionInstanceName)}
+	for _, name := range ruleList {
+		*securityGroupResp.SecurityRules = removeSecurityRuleBody(name, *securityGroupResp.SecurityRules)
+	}
+
+	err = createOrUpdateNetworkSecGroup(ctx, factory, opt, securityGroupResp)
+	if err != nil {
+		return err
+	}
+
+	logger.Info("bastion network security group rules removed: ", "nsg", NSGIngressAllowSSHResourceName(opt.BastionInstanceName), NSGEgressDenyAllResourceName(opt.BastionInstanceName), NSGEgressAllowOnlyResourceName(opt.BastionInstanceName))
+	return nil
+}
+
+func removePublicIP(ctx context.Context, factory azureclient.Factory, logger logr.Logger, opt *Options) error {
+	ip, err := getPublicIP(ctx, factory, opt)
+	if err != nil {
+		return err
+	}
+
+	if ip == nil {
+		return nil
+	}
+
+	publicClient, err := factory.NetworkPublicIP(ctx, opt.SecretReference)
+	if err != nil {
+		return err
+	}
+
+	err = publicClient.Delete(ctx, opt.ResourceGroupName, opt.BastionPublicIPName)
+	if err != nil {
+		return fmt.Errorf("failed to delete Public IP: %w", err)
+	}
+
+	logger.Info("Public IP removed", "ip", opt.BastionPublicIPName)
+	return nil
+}
+
+func removeNic(ctx context.Context, factory azureclient.Factory, logger logr.Logger, opt *Options) error {
+	nic, err := getNic(ctx, factory, opt)
+	if err != nil {
+		return err
+	}
+
+	if nic == nil {
+		return nil
+	}
```

Comment:
Do we need to fetch it upfront? Why not simply try to delete and return nil in case of not found error.

Reply:
ecf7ca13